qiskit/transpiler/passes/scheduling/dynamical_decoupling.py

```diff
@@ -153,102 +176,160 @@ def run(self, dag):
+        # Start = +16, End += 32
+        #
+        # final sequence       : 48dt-X1-80dt-Y2-80dt-X3-80dt-Y4-64dt / in total 992 dt
+        #
+        # Now we verify t0 of every node starts from multiple of 16 dt.
+        #
+        # X1:  48 dt (3 x 16 dt)
+        # Y2:  48 dt + 160 dt + 80 dt = 288 dt (18 x 16 dt)
+        # Y3: 288 dt + 160 dt + 80 dt = 528 dt (33 x 16 dt)
+        # Y4: 368 dt + 160 dt + 80 dt = 768 dt (48 x 16 dt)
+        #
+        # As you can see, constraints on t0 are all satified without explicit scheduling.
+        time_interval = t_end - t_start
+
+        if self._qubits and dag.qubits.index(qubit) not in self._qubits:
+            # Target physical qubit is not the target of this DD sequence.
+            delay_node = dag.apply_operation_back(Delay(time_interval, dag.unit), [qubit])
+            self._update_start_time(delay_node, t_start)
+            return
+
+        if self._skip_reset_qubits and (
+            isinstance(prev_node, DAGInNode) or isinstance(prev_node.op, Reset)
+        ):
+            # Previous node is the start edge or reset, i.e. qubit is ground state.
+            delay_node = dag.apply_operation_back(Delay(time_interval, dag.unit), [qubit])
+            self._update_start_time(delay_node, t_start)
+            return
+
+        slack = time_interval - np.sum(self._dd_sequence_lengths[qubit])
+        sequence_gphase = self._sequence_phase
+
+        if slack <= 0:
+            # Interval too short.
+            delay_node = dag.apply_operation_back(Delay(time_interval, dag.unit), [qubit])
+            self._update_start_time(delay_node, t_start)
+            return
+
+        if len(self._dd_sequence) == 1:
+            # Special case of using a single gate for DD
+            u_inv = self._dd_sequence[0].inverse().to_matrix()
+            theta, phi, lam, phase = OneQubitEulerDecomposer().angles_and_phase(u_inv)
+            if isinstance(next_node, DAGOpNode) and isinstance(next_node.op, (UGate, U3Gate)):
+                # Absorb the inverse into the successor (from left in circuit)
+                theta_r, phi_r, lam_r = next_node.op.params
+                next_node.op.params = Optimize1qGates.compose_u3(
+                    theta_r, phi_r, lam_r, theta, phi, lam
+                )
+                sequence_gphase += phase
+            elif isinstance(prev_node, DAGOpNode) and isinstance(prev_node.op, (UGate, U3Gate)):
+                # Absorb the inverse into the predecessor (from right in circuit)
+                theta_l, phi_l, lam_l = prev_node.op.params
+                prev_node.op.params = Optimize1qGates.compose_u3(
+                    theta, phi, lam, theta_l, phi_l, lam_l
+                )
+                sequence_gphase += phase
+            else:
+                # Don't do anything if there's no single-qubit gate to absorb the inverse
+                dag.apply_operation_back(Delay(time_interval, dag.unit), [qubit])
+                return
```

Comment:
Don't we need to call `_update_start_time()` here exceptionally? If we need to call it every time when we push back an op, how about introducing `BasePadding._apply_operation_back(dag, op, qubits, start, end)` ~=  `dag.apply_operation_back(op, qubits)` + `_update_start_time(start, end)` instead of `BasePadding._update_start_time()` to prevent from forgetting to call  `_update_start_time()`?

Reply:
Thanks @itoko ! That's neat. Done in https://github.com/Qiskit/qiskit-terra/pull/7760/commits/39ffd4e5348c95dd8ae903e388122b22b4de6b78